lesson6-Reflection/Task1/Container.cs

```
@@ -1,28 +1,100 @@
 ﻿using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
+using Task1.DoNotChange;
 
 namespace Task1
 {
     public class Container
     {
+        private readonly Dictionary<Type, Type> _mappedTypes = new Dictionary<Type, Type>();
+
         public void AddAssembly(Assembly assembly)
         {
-            throw new NotImplementedException();
+            var mappedTypes = assembly.GetTypes().Where(t =>
+                Attribute.IsDefined(t, typeof(ExportAttribute)) ||
+                Attribute.IsDefined(t, typeof(ImportConstructorAttribute)) ||
+                t.GetProperties().Any(p => Attribute.IsDefined(p, typeof(ImportAttribute))) ||
+                t.GetFields().Any(f => Attribute.IsDefined(f, typeof(ImportAttribute))));
+
+            foreach (var type in mappedTypes)
+                if (type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() == null)
+                    AddType(type);
+                else if ((type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() as ExportAttribute)?.Contract != null)
+                    AddType(type,
+                        (type.GetCustomAttributes(typeof(ExportAttribute)).First() as ExportAttribute)?.Contract);
+                else
+                    AddType(type);
         }
 
         public void AddType(Type type)
         {
-            throw new NotImplementedException();
+            // If Already added, update the injection type
+            if (_mappedTypes.ContainsKey(type))
```

Looking at the complex tests, if the type is already added and we try to add it again, it should throw an exception. So you should not do this check, it's enough just to add the type to dictionary.

---

done